account_move_line_purchase_info/migrations/16.0.2.0.0/post-migration.py

```diff
@@ -0,0 +1,19 @@
+from openupgradelib import openupgrade
+
+
+@openupgrade.migrate()
+def migrate(env, version):
+    purchase_orders = env["purchase.order"].search([("state", "=", "purchase")])
+    for order in purchase_orders:
+        valued_lines = order.order_line.invoice_lines.filtered(
+            lambda l: l.product_id
+            and l.product_id.cost_method != "standard"
+            and (not l.company_id.tax_lock_date or l.date > l.company_id.tax_lock_date)
+        )
+        svls, _amls = valued_lines._apply_price_difference()
+
+        if svls:
+            svls._validate_accounting_entries()
+
+        bills = order.invoice_ids.filtered(lambda bill: bill.state == "posted")
+        bills._stock_account_anglo_saxon_reconcile_valuation()
```

Comment:
@AlexPForgeFlow could you log the PO you are treating in each iteration? it can be helpful when tracking the logs.

We could also run the logic of the loop inside a try-except block in order to ensure that an upgrade is not blocked if there is any error trying to make this automatic correction.